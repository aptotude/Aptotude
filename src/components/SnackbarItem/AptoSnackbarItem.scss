.AptoSnackbarItem + .AptoSnackbarItem {
  margin-top: 0.5rem;
}

.AptoSnackbarItem {
  animation: show-snack 225ms;
  box-shadow: 0 3px 5px -1px rgba(0, 0, 0, 0.2),
    0 6px 10px 0 rgba(0, 0, 0, 0.14), 0 1px 18px 0 rgba(0, 0, 0, 0.12);
  &.AptoSnackbarItem--hide {
    animation: hide-snack 225ms;
    opacity: 0;
  }
}

.AptoSnackbarItemContent {
  display: flex;
  flex-wrap: nowrap;
  min-width: 288px;
  max-width: 568px;
  margin: 0;
}

.AptoSnackbarItemContentMessage {
  flex-grow: 1;
  max-width: 100%;
}

.AptoSnackbarItemContentAction {
  flex-grow: 0;
  padding-left: 1rem;
  max-width: 100%;
}

@keyframes show-snack {
  0% {
    transform: scale(0);
    opacity: 0;
  }
  100% {
    transform: scale(1);
    opacity: 1;
  }
}

@keyframes hide-snack {
  0% {
    transform: scale(1);
    opacity: 1;
  }
  100% {
    transform: scale(0);
    opacity: 0;
  }
}
