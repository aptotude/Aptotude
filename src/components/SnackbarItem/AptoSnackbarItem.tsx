import * as React from 'react';
import EventListener from 'react-event-listener';
import './AptoSnackbarItem.scss';
import { AptoAlertVariants, AptoAlert } from '../Alert';

interface AptoSnackbarItemProps {
  open?: boolean;
  action?: React.ReactNode;
  message?: React.ReactNode;
  variant?: AptoAlertVariants;
  autoHideDuration?: number;
  animationTimeout?: number;
  disableWindowBlurListener?: boolean;
  onMouseLeave?: (event: React.MouseEvent | null) => void;
  onMouseEnter?: (event: React.MouseEvent | null) => void;
  onClose?: (event: React.MouseEvent | null, type: string) => void;
  resumeHideDuration?: number;
}

interface AptoSnackbarItemState {
  exited: boolean;
  hiding: boolean;
}

export class AptoSnackbarItem extends React.Component<
  AptoSnackbarItemProps,
  AptoSnackbarItemState
> {
  public static defaultProps = {
    disableWindowBlurListener: false,
    variant: 'info',
    autoHideDuration: 5000,
    animationTimeout: 255
  };

  public state = {
    exited: false,
    hiding: false
  };

  public animationId = null;

  public timerAutoHide: any = null;

  public componentDidMount() {
    if (this.props.open) {
      this.setAutoHideTimer();
    }
  }

  public componentDidUpdate(prevProps: any) {
    if (prevProps.open !== this.props.open) {
      if (this.props.open) {
        this.setAutoHideTimer();
      } else {
        clearTimeout(this.timerAutoHide);
      }
    }
  }

  public componentWillUnmount() {
    clearTimeout(this.timerAutoHide);
  }

  public static getDerivedStateFromProps(nextProps: any, prevState: any) {
    if (nextProps.open) {
      return {
        exited: false
      };
    } else {
      return {
        exited: true
      };
    }
  }

  public setAutoHideTimer(autoHideDuration: number | null = null) {
    const autoHideDurationBefore =
      autoHideDuration != null ? autoHideDuration : this.props.autoHideDuration;

    if (!this.props.onClose || autoHideDurationBefore == null) {
      return;
    }

    clearTimeout(this.timerAutoHide);
    this.timerAutoHide = setTimeout(() => {
      const autoHideDurationAfter =
        autoHideDuration != null
          ? autoHideDuration
          : this.props.autoHideDuration;
      if (!this.props.onClose || autoHideDurationAfter == null) {
        return;
      }

      this.animateClose(null, 'timeout');
    }, autoHideDurationBefore);
  }

  public handleMouseEnter = (event: React.MouseEvent) => {
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(event);
    }
    this.handlePause();
  };

  public handleMouseLeave = (event: React.MouseEvent) => {
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(event);
    }
    this.handleResume();
  };

  public animateClose(event: any, type: string) {
    this.setState({
      hiding: true
    });

    setTimeout(() => {
      this.props.onClose && this.props.onClose(event, type);
    }, this.props.animationTimeout);
  }

  public handlePause = () => {
    clearTimeout(this.timerAutoHide);
  };

  public handleResume = () => {
    if (this.props.autoHideDuration != null) {
      if (this.props.resumeHideDuration != null) {
        this.setAutoHideTimer(this.props.resumeHideDuration);
        return;
      }
      this.setAutoHideTimer(this.props.autoHideDuration * 0.5);
    }
  };

  public handleExited = () => {
    this.setState({ exited: true });
  };

  public render() {
    const {
      variant,
      autoHideDuration,
      children,
      message,
      action,
      disableWindowBlurListener,
      onClose,
      onMouseEnter,
      onMouseLeave,
      open,
      animationTimeout,
      resumeHideDuration,
      ...rest
    } = this.props;

    if (!open && this.state.exited) {
      return null;
    }

    const hiding = this.state.hiding ? ' AptoSnackbarItem--hide' : '';

    return (
      <div
        className={`AptoSnackbarItem${hiding}`}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        {...rest}
      >
        <EventListener
          target="window"
          onFocus={disableWindowBlurListener ? undefined : this.handleResume}
          onBlur={disableWindowBlurListener ? undefined : this.handlePause}
        />
        <AptoAlert variant={variant} className="AptoSnackbarItemContent">
          {children || (
            <React.Fragment>
              <div className="AptoSnackbarItemContentMessage">{message}</div>
              <div className="AptoSnackbarItemContentAction">{action}</div>
            </React.Fragment>
          )}
        </AptoAlert>
      </div>
    );
  }
}
