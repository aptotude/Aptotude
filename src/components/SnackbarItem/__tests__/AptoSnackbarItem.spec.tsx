import React from 'react';
import { render, cleanup, wait, fireEvent } from 'react-testing-library';
import { AptoSnackbarItem } from '../AptoSnackbarItem';

describe('SnackbarItem Component', () => {
  afterEach(cleanup);

  it('It renders with message', () => {
    const props = {
      open: true,
      message: 'Hey'
    };
    const { container } = render(<AptoSnackbarItem {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('It renders null if closed', () => {
    const props = {
      open: false,
      message: 'Hey'
    };
    const { container } = render(<AptoSnackbarItem {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('It renders actions', () => {
    const props = {
      open: true,
      message: 'Hey',
      action: [<button key="foo">Foo</button>]
    };
    const { container } = render(<AptoSnackbarItem {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('It renders variant', () => {
    const props = {
      open: true,
      message: 'Hey',
      action: [<button key="foo">Foo</button>]
    };
    const { container } = render(
      <AptoSnackbarItem variant="danger" {...props} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('It eventually hides', async () => {
    const props = {
      open: true,
      onClose: jest.fn(),
      message: 'Hey',
      autoHideDuration: 0
    };
    const { container } = render(<AptoSnackbarItem {...props} />);
    await wait(() => {
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('It eventually calls onClose', async () => {
    const closer = jest.fn();
    const props = {
      open: true,
      onClose: closer,
      message: 'Hey',
      autoHideDuration: 0
    };
    render(<AptoSnackbarItem {...props} />);
    await wait(() => {
      expect(closer).toHaveBeenCalled();
    });
  });

  it('It does not hide if no onClose set', async () => {
    const props = {
      open: true,
      message: 'Hey',
      autoHideDuration: 0
    };
    const { container } = render(<AptoSnackbarItem {...props} />);
    await wait(() => expect(container.firstChild).toMatchSnapshot());
  });

  it('It does not call onClose if hovered over', async () => {
    const closer = jest.fn();
    const props = {
      open: true,
      onClose: closer,
      message: 'Hey',
      autoHideDuration: 50
    };
    const { container } = render(<AptoSnackbarItem {...props} />);

    fireEvent.mouseEnter(container);

    await wait(() => {
      expect(closer).not.toHaveBeenCalled();
    });
  });

  it('It eventually calls onClose when moused out', async () => {
    const closer = jest.fn();
    const props = {
      open: true,
      onClose: closer,
      message: 'Hey',
      autoHideDuration: 50
    };
    const { container } = render(<AptoSnackbarItem {...props} />);

    fireEvent.mouseEnter(container);
    fireEvent.mouseLeave(container);

    await wait(() => {
      expect(closer).toHaveBeenCalled();
    });
  });
});
