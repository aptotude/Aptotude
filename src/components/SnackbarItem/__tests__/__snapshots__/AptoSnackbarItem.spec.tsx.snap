// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackbarItem Component It does not hide if no onClose set 1`] = `
<div
  class="AptoSnackbarItem"
>
  <div
    class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
  >
    <div
      class="AptoSnackbarItemContentMessage"
    >
      Hey
    </div>
    <div
      class="AptoSnackbarItemContentAction"
    />
  </div>
</div>
`;

exports[`SnackbarItem Component It eventually hides 1`] = `
<div
  class="AptoSnackbarItem AptoSnackbarItem--hide"
>
  <div
    class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
  >
    <div
      class="AptoSnackbarItemContentMessage"
    >
      Hey
    </div>
    <div
      class="AptoSnackbarItemContentAction"
    />
  </div>
</div>
`;

exports[`SnackbarItem Component It renders actions 1`] = `
<div
  class="AptoSnackbarItem"
>
  <div
    class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
  >
    <div
      class="AptoSnackbarItemContentMessage"
    >
      Hey
    </div>
    <div
      class="AptoSnackbarItemContentAction"
    >
      <button>
        Foo
      </button>
    </div>
  </div>
</div>
`;

exports[`SnackbarItem Component It renders null if closed 1`] = `null`;

exports[`SnackbarItem Component It renders variant 1`] = `
<div
  class="AptoSnackbarItem"
>
  <div
    class="AptoAlert AptoAlert--danger AptoSnackbarItemContent"
  >
    <div
      class="AptoSnackbarItemContentMessage"
    >
      Hey
    </div>
    <div
      class="AptoSnackbarItemContentAction"
    >
      <button>
        Foo
      </button>
    </div>
  </div>
</div>
`;

exports[`SnackbarItem Component It renders with message 1`] = `
<div
  class="AptoSnackbarItem"
>
  <div
    class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
  >
    <div
      class="AptoSnackbarItemContentMessage"
    >
      Hey
    </div>
    <div
      class="AptoSnackbarItemContentAction"
    />
  </div>
</div>
`;
